src/main/scala/scorex/core/network/NodeViewSynchronizer.scala

```diff
@@ -214,65 +263,135 @@ class NodeViewSynchronizer[TX <: Transaction,
   }
 
   /**
-    * Logic to process modifiers got from another peer
+    * Logic to process modifiers got from another peer.
+    * Filter out non-requested modifiers (with a penalty to spamming peer),
+    * parse modifiers and send valid modifiers to NodeViewHolder
     */
   protected def modifiersFromRemote: Receive = {
     case DataFromPeer(spec, data: ModifiersData@unchecked, remote)
       if spec.messageCode == ModifiersSpec.MessageCode =>
 
       val typeId = data._1
       val modifiers = data._2
-
-      log.info(s"Got modifiers of type $typeId from remote connected peer: $remote")
-      log.trace(s"Received modifier ids ${data._2.keySet.map(encoder.encode).mkString(",")}")
-
-      modifiers.foreach { case (id, _) => deliveryTracker.onReceive(typeId, id, remote) }
-
-      val (spam, fm) = modifiers.partition { case (id, _) => deliveryTracker.isSpam(id) }
-
-      if (spam.nonEmpty) {
-        log.info(s"Spam attempt: peer $remote has sent a non-requested modifiers of type $typeId with ids" +
-          s": ${spam.keys.map(encoder.encode)}")
-        penalizeSpammingPeer(remote)
-        val mids = spam.keys.toSeq
-        deliveryTracker.deleteSpam(mids)
-      }
-
-      if (fm.nonEmpty) {
-        val mods = fm.values.toSeq
-        viewHolderRef ! ModifiersFromRemote(remote, typeId, mods)
+      log.info(s"Got ${modifiers.size} modifiers of type $typeId from remote connected peer: $remote")
+      log.trace(s"Received modifier ids ${modifiers.keySet.map(encoder.encode).mkString(",")}")
+
+      // filter out non-requested modifiers
+      val requestedModifiers = processSpam(remote, typeId, modifiers)
+
+      modifierSerializers.get(typeId) match {
+        case Some(serializer: Serializer[TX]@unchecked) if typeId == Transaction.ModifierTypeId =>
+          // parse all transactions and send them to node view holder
+          val parsed: Iterable[TX] = parseModifiers(requestedModifiers, serializer, remote)
+          viewHolderRef ! TransactionsFromRemote(parsed)
+
+        case Some(serializer: Serializer[PMOD]@unchecked) =>
+          // parse all modifiers and put them to modifiers cache
+          val parsed: Iterable[PMOD] = parseModifiers(requestedModifiers, serializer, remote)
+          parsed.foreach(pmod => putToCacheIfValid(remote, pmod))
```

Comment:
Sloppy code: if parse is empty, no any effects would be here, so it is logical to move this line after if() since you already has it